server/src/main/java/org/elasticsearch/action/SingleResultDeduplicator.java

```diff
@@ -59,10 +61,10 @@ public void execute(ActionListener<T> listener) {
                 return;
             }
         }
-        doExecute(listener);
+        doExecute(threadContext.newStoredContext(), ContextPreservingActionListener.wrapPreservingContext(listener, threadContext));
```

Comment:
Always wrap all listeners with `wrapPreservingContext`. I think this behaviour is more consistent in that when a listener is called, we can always expect it sees its original context. Right now, the first listener sees a changed context while the rest of the listeners see their own context even though work is done for all of them.